src/raglite/_rag.py

```diff
@@ -70,25 +78,197 @@ def create_rag_instruction(
+    chunks = []
+    clipped_messages = _clip(messages, max_tokens)
+    if tools and config.llm.startswith("llama-cpp-python"):
+        # Help llama.cpp LLMs plan their response by providing a JSON schema for the tool call.
+        clipped_messages[-1]["content"] += (
+            f"\n\nDecide whether to use or skip these tools in your response:\n{json.dumps(tools)}"
+        )
+    stream = completion(
+        model=config.llm,
+        messages=clipped_messages,
+        tools=tools,
+        tool_choice=tool_choice,
+        stream=True,
+    )
+    for chunk in stream:
+        chunks.append(chunk)
+        if isinstance(token := chunk.choices[0].delta.content, str):
+            yield token
+    # Check if there are tools to be called.
+    response = stream_chunk_builder(chunks, messages)
+    tool_calls = response.choices[0].message.tool_calls  # type: ignore[union-attr]
+    if tool_calls:
+        # Add the tool call request to the message array.
+        messages.append(response.choices[0].message.to_dict())  # type: ignore[arg-type,union-attr]
+        # Run the tool calls to retrieve the RAG context and append the output to the message array.
+        messages.extend(_run_tools(tool_calls, config))
+        # Stream the assistant response.
+        chunks = []
+        stream = completion(model=config.llm, messages=_clip(messages, max_tokens), stream=True)
+        for chunk in stream:
+            chunks.append(chunk)
+            if isinstance(token := chunk.choices[0].delta.content, str):
+                yield token
+    # Append the assistant response to the message array.
+    response = stream_chunk_builder(chunks, messages)
+    messages.append(response.choices[0].message.to_dict())  # type: ignore[arg-type,union-attr]
 
 
 async def async_rag(messages: list[dict[str, str]], *, config: RAGLiteConfig) -> AsyncIterator[str]:
-    # Truncate the oldest messages so we don't hit the context limit.
+    # If the final message does not contain RAG context, get a tool to search the knowledge base.
     max_tokens = get_context_size(config)
-    cum_tokens = np.cumsum([len(message.get("content", "")) // 3 for message in messages][::-1])
-    messages = messages[-np.searchsorted(cum_tokens, max_tokens) :]
-    # Asynchronously stream the LLM response.
-    async_stream = await acompletion(model=config.llm, messages=messages, stream=True)
-    async for output in async_stream:
-        token: str = output["choices"][0]["delta"].get("content") or ""
-        yield token
+    tools, tool_choice = _get_tools(messages, config)
+    # Asynchronously stream the LLM response, which is either a tool call or an assistant response.
+    chunks = []
+    clipped_messages = _clip(messages, max_tokens)
+    if tools and config.llm.startswith("llama-cpp-python"):
+        # Help llama.cpp LLMs plan their response by providing a JSON schema for the tool call.
+        clipped_messages[-1]["content"] += (
```

Comment:
Is this necessary? If I am not wrong, chattml-function-calling prompt format adds the tool schema to the system prompt automatically. 

https://github.com/abetlen/llama-cpp-python/blob/ea4d86ac25e0f39370ff3f54550387a02467ab77/llama_cpp/llama_chat_format.py#L3410

Reply:
It is also true that the `description` of the function seems missing, it looks as an overlook.

```python
        "{% for tool in tools %}"
        "\nfunctions.{{ tool.function.name }}:\n"
        "{{ tool.function.parameters | tojson }}"
        "\n{% endfor %}"
```